execute/tokendata/observer_background.go

```diff
@@ -0,0 +1,436 @@
+package tokendata
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	mapset "github.com/deckarep/golang-set/v2"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+
+	"github.com/smartcontractkit/chainlink-ccip/execute/exectypes"
+	cciptypes "github.com/smartcontractkit/chainlink-ccip/pkg/types/ccipocr3"
+)
+
+type backgroundObserver struct {
+	lggr              logger.Logger
+	observer          TokenDataObserver
+	numWorkers        int
+	cachedTokenData   *inMemTokenDataCache
+	msgQueue          *msgQueue
+	wg                sync.WaitGroup
+	done              chan struct{}
+	observeTimeout    time.Duration
+	reprocessInterval time.Duration // how long to wait before reprocessing a message
+}
+
+// NewBackgroundObserver initializes an observer that retrieves and caches token data in the background.
+// It uses the provided observer to make the actual Observe calls, storing results in memory for efficient access later.
+// Goroutines are spawned to process messages concurrently, numWorkers defines how many.
+// cacheExpirationInterval defines for how long in memory token data are considered active.
+// cacheCleanupInterval defines how often to check and cleanup inactive data.
+// observeTimeout defines how long to wait for the underlying observer to return results.
+func NewBackgroundObserver(
+	lggr logger.Logger,
+	observer TokenDataObserver,
+	numWorkers int,
+	cacheExpirationInterval time.Duration,
+	cacheCleanupInterval time.Duration,
+	observeTimeout time.Duration,
+) TokenDataObserver {
+	doneChan := make(chan struct{})
+
+	o := &backgroundObserver{
+		lggr:       lggr,
+		observer:   observer,
+		numWorkers: numWorkers,
+		cachedTokenData: newInMemObservationsCache(
+			logger.Named(lggr, "inMemObservationsCache"),
+			cacheExpirationInterval,
+			cacheCleanupInterval,
+			doneChan,
+		),
+		msgQueue:          newMsgQueue(logger.Named(lggr, "msgQueue")),
+		wg:                sync.WaitGroup{},
+		done:              doneChan,
+		observeTimeout:    observeTimeout,
+		reprocessInterval: 5 * time.Second,
+	}
+
+	o.startWorkers()
+	return o
+}
+
+// Observe fetches token data for the given messages that are already present in memory.
+// If token data are not in memory, it enqueues the messages for background processing.
+// Meaning that at least two calls to Observe are needed to get token data for a message.
+// If data are already in the queue for processing, it is a nop.
+func (o *backgroundObserver) Observe(
```

Comment:
Observe is probably the wrong name. This background process is fetching attestations as quickly as possible, it shouldn't worry about what they're used for. Maybe something like `GetCachedTokenData` or `GetAvailableTokenData`?